qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+
+        .. note:: All parametrized object should be stored in the operands.
+            Otherwise parameter cannot be detected.
+        """
+        node._operands = tuple(self.visit(op) for op in node.operands)
+
+        return node
+
+    # Lower layer: Assign parameters to operands
+
+    def visit_Channel(self, node: channels.Channel):
+        """Assign parameters to ``Channel`` object."""
+        if isinstance(node.index, ParameterExpression):
+            new_index = self._assign_parameter_expression(node.index)
+
+            # validate
+            if not isinstance(new_index, ParameterExpression):
+                if not isinstance(new_index, int) or new_index < 0:
+                    raise PulseError('Channel index must be a nonnegative integer')
+
+            # return new instance to prevent accidentally override timeslots without evaluation
+            return node.__class__(index=new_index)
+
+        return node
+
+    def visit_ParametricPulse(self, node: ParametricPulse):
+        """Assign parameters to ``ParametricPulse`` object."""
+        new_parameters = {}
+        for op, op_value in node.parameters.items():
+            if isinstance(op_value, ParameterExpression):
+                op_value = self._assign_parameter_expression(op_value)
+            new_parameters[op] = op_value
+
+        return node.__class__(**new_parameters, name=node.name)
+
+    def visit_Waveform(self, node: Waveform):
+        """Assign parameters to ``Waveform`` object.
+
+        .. node:: No parameter can be assigned to ``Waveform`` object.
+        """
+        return node
+
+    def generic_visit(self, node: Any):
+        """Assign parameters to object that doesn't belong to Qiskit Pulse module."""
+        if isinstance(node, ParameterExpression):
+            return self._assign_parameter_expression(node)
+        else:
+            return node
+
+    def _assign_parameter_expression(self, param_expr: ParameterExpression):
+        """A helper function to assign parameter value to parameter expression."""
+        new_value = copy(param_expr)
+        for parameter in param_expr.parameters:
+            if parameter in self._param_map:
+                new_value = new_value.assign(parameter, self._param_map[parameter])
+
+        return format_parameter_value(new_value)
+
+    def _update_parameter_manager(self, node: Union[Schedule, ScheduleBlock]):
+        """A helper function to update parameter manager of pulse program."""
+        new_parameters = set()
+
+        for parameter in node.parameters:
+            if parameter in self._param_map:
+                value = self._param_map[parameter]
+                if isinstance(value, ParameterExpression):
+                    for new_parameter in value.parameters:
+                        new_parameters.add(new_parameter)
+            else:
+                new_parameters.add(parameter)
+        node._parameter_manager._parameters = new_parameters
```

Comment:
Do you have to check that `node` has a parameter manager? Why is this attribute private?

Reply:
Yes. This helper function is called only by programs, i.e. schedule and blocks. Because schedules can be nested, nested schedule says parameter is unassigned until parameter manager is updated. Schedules and blocks don't manage parameters by itself.